schemas/fingerprint-server-api.yaml

```diff
@@ -1931,3 +1967,32 @@ components:
           required:
             - code
             - message
+    VelocityIntervals:
+      type: object
+      additionalProperties: false
+      properties:
+        'intervals':
+          $ref: '#/components/schemas/VelocityIntervalResult'
+    VelocityIntervalResult:
+      type: object
+      description: >
+        Is absent if the velocity data could not be generated for the visitor ID.
+      additionalProperties: false
+      properties:
+        '5m':
+          type: integer
+          example: 1
+        '1h':
+          type: integer
+          example: 1
+        '24h':
+          type: integer
+          description: >
+            In some cases 24h interval could be absent if the visitor has extremely high activity
+            (which confidently signifies automation activity) thereby 24h window doesn't make sense as the numbers would be
+            too high to make any relevance
```

Comment:
So if a visitor is in high-activity for 20 hours, spamming thousands of requests, then stops for an hour and then starts again, it will receive the following Velocity signals? 

```json
{ "5m": 1, "1h": 1 }  
```

And the way to detect such a high-activity user is to check 

```js
if (velocity.events.intervals['24h'] === undefined) { 
  console.log("Extremely high activity user") 
} 
```
?

If you you didn't carefully read the docs, wouldn't you assume that a missing `24h` interval just means the user didn't have any requests beyond the provided `1h`? Wouldn't returning the actual `24h` value, or some documented maximum value or "Infinity" or something like that be more intuitive/semantic?

Reply:
Yes, basically we would be updating in the documentation a benchmark number suppose more than 20k requests in a day (would mean high activity) and this would ignore aggregation of results for the larger window as its computationally challenging

Reply:
Also we would be providing the `events` **24h** count because its a simple counter but for other signals like unique Ip's, linked Id's this would be not possible

Reply:
> Yes, basically we would be updating in the documentation a benchmark number suppose more than 20k requests in a day (would mean high activity) and this would ignore aggregation of results for the larger window as its computationally challenging 

I am not sure I understand, can you rephrase? 

My argument is that the way it's currently working is confusing. It's not obvious from the response itself that this is a high-activity user and it should be. If 20k is our limit, then we should return 

`{ "5m": 1, "1h": 1, "24h": 20000 }` 

And document that the 20k could actually be more, but we only count up to 20k. But we should show the user 20k, or whatever the limit is, not `undefined`.

Reply:
@ilfa what do you think?

Reply:
I agree, current behavior is a little bit confusing.

Reply:
Let me provide you with some details,

1. We have 4 Velocity Signals as of now
2. Out of these 4 only `events` type is a simple counter
3. So as the example you provided above this Velocity signal will always provide us results for 5m, 1hr and 24hr window
4. but for the other 3 signals, It becomes computationally costly to count and compute distinct and thereby we ignore the calculation for 24 hour window as this increases the latency by almost 10x.

Let me know If this clear your doubts

Reply:
@ilfa , @JuroUhlar Let me know the best possible place to document this limitation (as we would be sticking with this solution for a while)

Reply:
So to be clear: 

* `events` will **always** return a `24h` value 
* the other velocity signals will **never** return a `24h` value 

Correct?

Reply:
By never only in the case events 24h will have more than 20k events

Reply:
Ok so 

* `events` will return a `24h` value if that value is below `20000` and `undefined` if it's above `20000` 
* `distinctIp`, `distinctLinkedId`, and `distinctCountry` will **never** return a `24h` value

Is that accurate? If so, and you are not willing to change it, I think it needs to be spelled out in the documentation clearly like above.

If you are willing to change it, I think the `events` behavior should change to return `20000` instead of `undefined` for all values above `20000`.

We should consider splitting it into two schema components

* A `ShortIntervals` like `{ '1h': number, '5m': number }` 
* A `ExtendedIntervals` like `{ '1h': number, '5m': number, '24h': number }` with a note that 24h only counts up to 20k. 

Then assign the appropriate schema component to the appropriate signal

Reply:
No
What I meant is `events` will **always** have all the 3 windows 

Now regarding other signals if the `events` `24h` window crosses the 20k mark we will skip the `24h`window for these remaining 3 Signals(LinkedId, Ip, Country)

Now from your comment it's clear to me about the detailed documentation you want around this(I will be adding all the information soon)

But can you help me explain and understand the change where you expect them to be broken down into extended and normal ? Is this actually important or just detailing in depth will suffice

Reply:
Aah, okay, I think I got it now. Let us confirm: 

If a visitor generates 20k events using 100 different IP addresses in a 24h period, on the 20000th event, they will receive the following response 

```json5
{
  "data": {
    // Same for countries, linkedId
    "distinctIp": {
      "intervals": {
        "5m": 10,
        "1h": 20,
        "24h": 100
      }
    },
    "events": {
      "intervals": {
        "5m": 1000,
        "1h": 2000,
        "24h": 20000
      }
    }
  }
}
```

But on the next, 20001st event, it will receive this, with the 24h value missing for intervals other than `events`, correct? 

```json5
{
  "data": {
    // Same for countries, linkedId
    "distinctIp": {
      "intervals": {
        "5m": 11,
        "1h": 21,
      }
    },
    "events": {
      "intervals": {
        "5m": 1001,
        "1h": 2001,
        "24h": 20001
      }
    }
  }
}
```

Is that correct?

If my understanding is now correct, I still think we should consider changing the implementation. If I were a user looking at the second response, I would still rather receive the last available value for `distinctIp.24h`, even if it only included IPs for the last 20k events and not necessarily the entire 24h period. It's still useful to roughly know if it was 2 IP addresses or a 2000. Right now, to find this information, I would need to make hundreds/thousands of Server API calls to find the 20000th event that had the info. We can then document it as 

> If the `events.intervals.24h` count exceeds 20.000 events, the `24h` interval of other velocity signals in the response like `distinctIp.intervals.24h` will only reflect the most recent 20.000 events." 

What do you think? 
If it's not possible, use the sentence to document the current state: 

> If the `events.intervals.24h` count exceeds 20.000 events, the `24h` interval of all other velocity signals in the response will be omitted".

> But can you help me explain and understand the change where you expect them to be broken down into extended and short ?

If my current understanding is correct, it might not be strictly necessary, but it would be slightly more accurate to say that `events.intervals.24h` is of type 'FullIntervals = { 5m: number, 1h: number, 24: number }` and all other signals are of type `PartialIntervals =  { 5m: number, 1h: number, 24: number | undefined }`. @ilfa What do you think?

Reply:
Thank you for the detailed explanation. Your understanding is correct. @JuroUhlar 

For events exceeding the 20k threshold within a 24h window, the response will indeed omit the 24h interval for other signals like distinctIp, LinkedId, and Country from the 20001st event onwards. 

When considering the options for handling this, we had a few in mind:

1. Writing it as "24h": -1
2. Omitting the 24h interval

We decided to go with the second approach because displaying a placeholder value such as -1, 0, or 2000 does not accurately reflect the actual count. Additionally, we have implemented a similar feature where we hide the flag of dailyRequests if it’s false in a high-activity device signal so we wanted to stay consistent.

Reply:
If we are okay can I implement the changes as suggested by you by breaking the intervals into 2 types partial and complete ? @JuroUhlar

Reply:
Thank you! I still think that including the imperfect value would be better DX than omitting it, but it's ultimately a product decision to weigh the trade-offs.

I pushed a commit adjusting the explanation copy. 

@ilfa What do you think about the two intervals? Useful or overkill? I don't have a strong opinion.

Reply:
To clarify: if there were no events, we would get `{ 5m: 0, 1h: 0, 24: 0 }`.
If there were 20K+ events 5 hours ago and no one in the last hour we will get `{ 5m: 0, 1h: 0 }`.

In this case and Juraj's explanation it looks ok in DX terms (it is easy to identify such an event just by checking only one field).

Reply:
No @ilfa , it's an atomic operation to increment the counters and ideally the response is always generated afterwards an valid event so we will always have non 0 value for all the three
(So if you are initiating the request for the first time you will get 1,1,1 for 5m,1h,24h)

And for the other case where meta data is not sent by client for example linkedID we will not have anything inside the key 
So `interval` is an optional member

Reply:
Hi team, we discussed this internally and found that there are pros and cons to both options (omitting and leaving it as 2000). We can go with 2000, as it does provide some value to the customer. Thanks!

Reply:
I don't think it's technically possible, because the whole purpose is to drop aggregation for high traffic visitors
As they block the whole pipeline (and we cannot add any bit of latency to existing warden execution)

Reply:
@soniabhishek synced offline, we agree on omitting the "24h" threshold

Reply:
I agree that if providing even the approximate 24h value carries latency risk (which it seems it does) then it makes sense to omit the 24h value despite the worse DX —  as long as the behavior is clearly documented (which it seems it now is). Thank you, resolving!